polkadot/node/service/src/tests.rs

```diff
@@ -243,7 +243,7 @@ impl ChainBuilder {
 			parent_hash != Self::GENESIS_PARENT_HASH,
 			"cannot add block with genesis parent hash"
 		);
-		assert!(self.0.blocks_by_hash.len() < u8::MAX.into());
+		assert!(self.0.blocks_by_hash.len() < u8::MAX as usize);
```

Comment:
I have a deep hatred of the `as` operator (as it can result in extremely hard to find catastrophic bugs after a seemingly innocuous refactor; ask me how I know), so this seems like a step backwards. Why not `usize::from(u8::MAX)`?

Reply:
Agreed. Should use `From` or `TryFrom`. `as` only if absolutely necessary.

Reply:
totally agree, safety overlooked. Adjusted some of these unsafe cast here: [adf6ac6](https://github.com/paritytech/polkadot-sdk/pull/8118/commits/adf6ac6454cff1bf7fac1cccd12954e742a7a112). There are still some places with coercions like `vec![] as Vec<T>`, but I don't find them risky

Reply:
Ok, now there are a few errors related to missing `usize::from` impls. `try_from` can be used but we need to handle errors in some non-test related places. I'll have a new iteration with this handling done where it makes sense.

Reply:
> There are still some places with coercions like `vec![] as Vec<T>`, but I don't find them risky

Yes those are not risky. They are basically just helping type inference. However, I would still prefer to avoid the `as` syntax where possible. It is just mental load when reading `as` to figure out of this is an okay use case. Rather use type annotations than as.

Reply:
@athei @koute what do you think about handling the error here vs `expect`ing: https://github.com/paritytech/polkadot-sdk/pull/8118/commits/39803747ac9ab36e0efdfd32c936f263457e93cb? I doubt the code will ever run on platforms where sizeof(`usize`) < sizeof(`u32`), or the fields will hold negatives, so expecting can also make sense, while adding appropiate comments. I am asking mainly because if we handle the error my way, then we need a major bump for `polkadot-runtime-parachains` (I introduce a new error variant), which means this PR can not be backported to stable2412/stable2503.

Reply:
While I agree that is it most likely never hit it is still best practice to do so. What if some point we bump the `u32` to `u64` and then it silently passes? So I would say keep using `TryFrom` here.

> I am asking mainly because if we handle the error my way, then we need a major bump for polkadot-runtime-parachains

Just use the existing `HeadDataTooLarge`. If the value is too large to fit into an array it is too large :) We also don't need a log here I think. The error is quite easy to pin point.

Reply:
@iulianbarbu Even though Rust doesn't have an `Into`/`From` impl for this you can safely assume that `size_of::<usize>() >= size_of::<u32>()`; we will never run on any platform where `usize` is not at least as big as `u32`, so you can `unwrap`/`expect` those conversions (but only if the types are *explicit*).

Reply:
Yes. In this case we don't need to use `expect` (even if we could) because the function allows for returning errors anyways. We mostly want to avoid silent truncation with `as`.

Reply:
Addressed everything in this last commit: [5a1274f](https://github.com/paritytech/polkadot-sdk/pull/8118/commits/5a1274f9ab49cd16d84aaa2c4a92d3a5de274ba4).